pyfar/dsp/dsp.py

```diff
@@ -1386,6 +1386,241 @@ def time_shift(signal, shift, unit='samples'):
+        >>> start_samples = pf.dsp.find_impulse_response_start(ir)
+        >>> ax = pf.plot.time(ir, unit='ms', label='impulse response', dB=True)
+        >>> ax.axvline(
+        ...     start_samples/ir.sampling_rate*1e3,
+        ...     color='k', linestyle='-.', label='start sample')
+        >>> ax.axhline(
+        ...     20*np.log10(np.max(np.abs(ir.time)))-20,
+        ...     color='k', linestyle=':', label='threshold')
+        >>> ax.legend()
+
+    Create a train of weighted impulses with levels below and above the
+    threshold, serving as a very abstract room impulse response. The starting
+    sample is identified as the last sample below the threshold relative to the
+    maximum of the impulse response.
+
+    .. plot::
+
+        >>> import pyfar as pf
+        >>> import numpy as np
+        >>> n_samples = 64
+        >>> delays = np.array([14, 22, 26, 30, 33])
+        >>> amplitudes = np.array([-35, -22, -6, 0, -9], dtype=float)
+        >>> ir = pf.signals.impulse(n_samples, delays, 10**(amplitudes/20))
+        >>> ir.time = np.sum(ir.time, axis=0)
+        >>> start_sample_est = pf.dsp.find_impulse_response_start(
+        ...     ir, threshold=20)
+        >>> ax = pf.plot.time(
+        ...     ir, dB=True, unit='samples',
+        ...     label=f'peak samples: {delays}')
+        >>> ax.axvline(
+        ...     start_sample_est, linestyle='-.', color='k',
+        ...     label=f'ir start sample: {start_sample_est}')
+        >>> ax.axhline(
+        ...     20*np.log10(np.max(np.abs(ir.time)))-20,
+        ...     color='k', linestyle=':', label='threshold')
+        >>> ax.legend()
+
+    """
+    ir_squared = np.abs(impulse_response.time)**2
+
+    mask_start = np.int(0.9*impulse_response.n_samples)
+
+    mask = np.arange(mask_start, ir_squared.shape[-1])
+    noise = np.mean(np.take(ir_squared, mask, axis=-1), axis=-1)
+
+    max_sample = np.argmax(ir_squared, axis=-1)
+    max_value = np.max(ir_squared, axis=-1)
+
+    if np.any(max_value < 10**(threshold/10) * noise) or \
+            np.any(max_sample > mask_start):
+        warnings.warn(
+            "The SNR seems lower than the specified threshold value. Check "
+            "if this is a valid impulse response with sufficient SNR.")
+
+    start_sample = max_sample.copy()
+
+    for ch in np.ndindex(impulse_response.cshape):
+        # Only look for the start sample if the maximum index is bigger than 0
+        if start_sample[ch] > 0:
+            # Check samples before maximum
+            ir_before_max = np.squeeze(
+                ir_squared[ch][:max_sample[ch]+1] / max_value[ch])
+            # First sample above or at the threshold level
+            idx_first_above_thresh = np.where(
+                ir_before_max >= 10**(-threshold/10))[0]
+            if idx_first_above_thresh.size > 0:
+                # The start sample is the last sample below the threshold
+                start_sample[ch] = np.min(idx_first_above_thresh) - 1
```

Comment:
Could we use `idx_first_above_thresh[0]` instead? I think the result is already sorted

Reply:
It is, but this is fail save ;)